src/tribler-core/tribler_core/modules/popularity/test_popularity_community.py

```diff
@@ -75,6 +75,51 @@ async def test_torrents_health_gossip(self):
             assert torrent.leechers == checked_torrent_info[2]
             assert torrent.last_check == checked_torrent_info[3]
 
+    async def test_torrents_health_gossip_multiple(self):
+        """
+        Test whether torrent health information is correctly gossiped around
+        """
+        # torrent structure is (infohash, seeders, leechers, last_check)
+        dead_torrents = {(random_infohash(), 0, randint(1, 10), int(time.time()))
+                         for _ in range(10)}
+        alive_torrents = {(random_infohash(), randint(1, 10), randint(1, 10), int(time.time()))
+                          for _ in range(PopularityCommunity.GOSSIP_RANDOM_TORRENT_COUNT)}
+        top_popular_torrents = {(random_infohash(), randint(11, 100), randint(1, 10), int(time.time()))
+                                for _ in range(PopularityCommunity.GOSSIP_POPULAR_TORRENT_COUNT)}
+
+        all_checked_torrents = dead_torrents | alive_torrents | top_popular_torrents
+
+        db1 = self.nodes[0].overlay.mds.TorrentState
+        db2 = self.nodes[1].overlay.mds.TorrentState
+
+        with db_session:
+            assert db1.select().count() == 0
+            assert db2.select().count() == 0
+
+        for torrent_info in all_checked_torrents:
+            self.nodes[0].overlay.torrent_checker.torrents_checked.add(torrent_info)
+
+        await self.introduce_nodes()
+
+        # Gossip random torrents
+        self.nodes[0].overlay.gossip_random_torrents_health()
+        await self.deliver_messages(timeout=0.1)
+
+        # Check whether node 1 has received all random torrent health information
+        with db_session:
+            assert db2.select().count() == PopularityCommunity.GOSSIP_RANDOM_TORRENT_COUNT
+
+        # Gossip popular torrents
+        self.nodes[0].overlay.gossip_popular_torrents_health()
+        await self.deliver_messages(timeout=0.1)
+
+        # Check whether node 1 has received all popular torrent health information.
+        with db_session:
+            count = db2.select().count()
+            assert count >= PopularityCommunity.GOSSIP_POPULAR_TORRENT_COUNT
+            assert count <= PopularityCommunity.GOSSIP_RANDOM_TORRENT_COUNT \
```

Comment:
The count should be `==` here since you completely control the environment, and there is no reinsertion in sampling the torrents.

Reply:
Random torrents could include popular torrents as well so there is this check for range instead of `==`

Reply:
Hmm... :thinking: 
https://github.com/Tribler/tribler/blob/3c2abfbad404fb52fe756ec830d86335b6554009/src/tribler-core/tribler_core/modules/popularity/popularity_community.py#L77

This line suggests that you're sampling `random` from the `rest` which is, well, the set of  `all` with removed `popular`, right? That is basically sampling without reinsertion.

Even if you're doing reinsertion, putting the test like you did, with `a<=b<=c`, may still result in `a==b` always. E.g. the test will **still pass** even if you disable sampling random torrents, right? That is clearly a deficiency.

BTW, to avoid randomness you can use `random.seed(123)` or something like that. You can achieve it by, say, `Patch`ing the `random` module in the community namespace. For reference, see:
https://github.com/Tribler/tribler/blob/74f836c4075b592597529438c734d99d683eb3c1/src/tribler-core/tribler_core/modules/metadata_store/tests/test_channel_metadata.py#L506
https://github.com/Tribler/tribler/blob/74f836c4075b592597529438c734d99d683eb3c1/src/tribler-core/tribler_core/modules/metadata_store/community/tests/test_gigachannel_community.py#L49

Reply:
Yes, I'm sampling from the `rest` but random sampling will still include popular torrents because in computing random torrents `include_popular=False` is passed so sampling happens in the whole instead of `rest`. This is still important because it introduces the random uniform probability of including popular torrent in the first 2 mins because an explicit popular torrent message is gossiped.

```python
   def gossip_random_torrents_health(self):
        """
        Gossip random torrent health information to another peer.
        """
        self._gossip_torrents_health(include_popular=False, include_random=True)

    def gossip_popular_torrents_health(self):
        """
        Gossip popular torrent health information to another peer.
        """
        self._gossip_torrents_health(include_popular=True, include_random=False)
```

Regarding the test, I've updated to check the overlap and the existence of the popular torrents.

Reply:
That's some confusing usage of names :confused:
Anyways, let's merge it as it is. I'll fix it later today.